### PR TITLE
Support boost with only subset of libs

### PR DESCRIPTION
**Pull request by helmesjo**
Hi, and _thanks first and foremost for making my life a lot easier_!

I've written a [boost-download](https://github.com/helmesjo/hello-cd/blob/0bf83f72685a1244914faccf83f5d8836ac61576/src/cmake/boost-download.cmake) script which automates the download of boost, and after pulling all submodules it builds and run [bcp
](http://www.boost.org/doc/libs/1_65_1/tools/bcp/doc/html/index.html) to extract the minimum required subset of boost for the modules specified as arguments.

A lot of details, but in the end a trimmed down boost-repo is merged into my repo (~50mb instead of 500+ with shallow clone & pull).

So, with this little change to your wonderful repo I can properly build even though not all modules are part of the BOOST_SOURCES.

(I preferred declaring two explicit lists instead of if:ing on "headers" also)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -60,10 +61,17 @@ set(BOOST_LIBS
...
 
-foreach(lib ${BOOST_LIBS})
+foreach(lib ${BOOST_LIBS_REQUIRED})
   include("libs/${lib}.cmake")
 endforeach()
 
+foreach(lib ${BOOST_LIBS_OPTIONAL})
+  # In case only a subset of modules is available (eg. after using bcp)
+  if(EXISTS "${BOOST_SOURCE}/libs/${lib}")
+    include("libs/${lib}.cmake")
```
Comment (Orphis) [inline CMakeLists.txt:71]: It could as well be: `include("libs/${lib}.cmake" OPTIONAL)`

**Comment by Orphis** [comment]
Hej!
I'm glad you find this project useful!

This is a great idea and we should definitely support this scenario. I believe just adding `OPTIONAL` to the `include()` should be enough though.
Also, I'm not sure about the build failures, could you have a look?

**Comment by helmesjo** [comment]
Aha, didn't know about that flag! That is indeed much prettier.
Yeah I  noticed, but had to give in for my tired eyes... I'll have a look at it tonight!

**Comment by helmesjo** [comment]
Oh btw, just realized... Optional won't do since whether it should be included or not is determined by the source dir specifically (if the actual module exists), not boost-cmake/libs/*.cmake (which is where the OPTIONAL is applied).

I'll still have to look at the build errors though!

**Comment by Orphis** [comment]
Doh, right, that's what happens when you do post midnight reviews...
The problem for the test failure is that the CMake library name doesn't match the folder name.
I want to be compatible with FindBoost.cmake, so we can't rename the library but we can rename the file in lib/ to match probably.

**Comment by helmesjo** [comment]
Yep, peeked at it this morning and indeed it is. 
Do you mean renaming the *.cmake files in /lib, say `lib/unit_test_framework.cmake`->`lib/test.cmake` (but keep the target name as-is)? That should work, and make the mapping from git module<->cmake-file more intuitive.

How come the mismatch with test<->unit_test_framework to begin with? Namespace used by boost is boost::unit_test_framework, but module in git-repo was named "test"...

Anyways, I'll try your suggestion tonight!

**Comment by helmesjo** [comment]
Any chance you could merge this, and perhaps create a new release..?

File: cmake/Modules/AddBoostLib.cmake [hunk 2]
```diff
@@ -34,4 +34,6 @@ function(_add_boost_lib)
   if(BOOSTLIB_INCLUDE_PRIVATE)
     target_include_directories(Boost_${BOOSTLIB_NAME} PRIVATE ${BOOSTLIB_INCLUDE_PRIVATE})
   endif()
+
+  message("Added Boost::${BOOSTLIB_NAME}")
```
Comment (Orphis) [inline cmake/Modules/AddBoostLib.cmake]: Could you remove this bit? I'm afraid it might be a bit spammy.

**Review by Orphis — CHANGES_REQUESTED** [review]
Can you add the requested change and squash the commits please?

**Comment by Orphis** [comment]
And yes, I'll merge the change soon. I should start working on integrating 1.65.1 too.
Is that what you've been using maybe?

File: cmake/Modules/AddBoostLib.cmake (hunk 2 above)
Comment (helmesjo) [reply cmake/Modules/AddBoostLib.cmake]: Sure! Was mostly just to verify the behavior.

**Comment by helmesjo** [comment]
Currently running 1.64.0, but 1.65.1 should work just as well! I'll fix the change + squash tonight (probably).

About verification, should any additional tests be added? Perhaps (at least) a build verifying that some predetermined minimal subset of boost can be built... 

**Example:**  
1. Download boost.
2. Remove all but a subset of the modules.
3. Build.

**Review by helmesjo — COMMENTED** [review]
Removed logging & squashed commits.

**Review by Orphis — APPROVED** [review]

**Pull request merged**